test/models/product_test.rb

```diff
@@ -0,0 +1,200 @@
+      #
+      # prod = products(:tricycle)
+      # order = orders(:order2)
+      # order_prod = OrderProduct.create!(quantity: 1, product: prod, order: order)
+      #
+      # prod.order_products.product_id.must_equal order.order_products.product_id
+    end
+
+    it "can have reviews" do
+      @p.must_respond_to :reviews
+
+      product = products(:tricycle)
+
+      review = Review.create!(rating: 4, text: "great", product_id: product.id)
+      product.reviews << review
+      product.reviews.must_include review
+    end
+
+    it "has at least one category, and can have multiple" do
+      @p.must_respond_to :categories
+
+      @p.categories.must_be :empty?
+
+      cat = Category.create!(name: "computers")
+      @p.categories << cat
+      @p.categories.must_include cat
+
+      cat2 = Category.create!(name: "three-legged chairs")
+      @p.categories << cat2
+      @p.categories.must_include cat2
+    end
+
+  end
+
+  describe 'validations' do
+    it 'is valid' do
+      b = Product.new(name: "book", price: 2, merchant_id: Product.first.merchant_id, inventory: Product.first.inventory)
+      b.category_ids = Category.first.id
+      b.must_be :valid?
+    end
+
+    it 'requires a name' do
+      is_valid = @p.valid?
+      is_valid.must_equal false
+      @p.errors.messages.must_include :name
+    end
+
+    it 'requires a unique name' do
+      name = "test book"
+      Product.create!(name: name, price: 1, merchant_id: Product.first.merchant_id, inventory: Product.first.inventory, category_ids: Category.first.id)
+      b2 = Product.new(name: name, price: 2, merchant_id: Product.first.merchant_id, inventory: Product.first.inventory, category_ids: Category.first.id)
+      b2.wont_be :valid?
+    end
+
+    it 'requires a price' do
+      is_valid = @p.valid?
+      is_valid.must_equal false
+      @p.errors.messages.must_include :price
+    end
+
+    it 'requires a price greater than 0' do
+      b = Product.new(name: "tom petty video", price: -1)
+      is_valid = b.valid?
+      is_valid.must_equal false
+      b.errors.messages.must_include :price
+    end
+
+    it "requires inventory that is greater or equal to 0" do
+      p = Product.new(name: "tom petty video", price: 1)
+      is_valid = p.valid?
+      is_valid.must_equal false
+      p.errors.messages.must_include :inventory
+
+      p2 = Product.new(name: "tom petty video", price: 1, inventory: -1)
+      is_valid = p2.valid?
+      is_valid.must_equal false
+      p2.errors.messages.must_include :inventory
+    end
+
+    it "requires the presence of a category" do
+      p = Product.new(name: "tom petty video", price: 1, inventory: 1)
+      is_valid = p.valid?
+      is_valid.must_equal false
+      p.errors.messages.must_include :category_ids
+    end
+  end
+
+  #TODO:
+  describe "increase_inventory" do
+
+
+  end
+
+  describe "decrease_inventory" do
+    let(:tricycle) { products(:tricycle) }
+    it "decreases the inventory by the given quantity" do
```

Comment:
What if you try to decrease the inventory by 0, or a negative number? What if you try to decrease it past the number in stock?